File: Strategy/Java/src/duck/Duck.java
package eu.svetlin.design.patterns.strategy.duck;

import eu.svetlin.design.patterns.strategy.fly.FlyBehavior;
import eu.svetlin.design.patterns.strategy.quack.QuackBehavior;

public abstract class Duck {
    FlyBehavior flyBehavior;
    QuackBehavior quackBehavior;

    public Duck() {}

    public abstract void display();

    public void performFly() {
        flyBehavior.fly();
    }

    public void setFlyBehavior(FlyBehavior flyBehavior){
        this.flyBehavior = flyBehavior;
    }

    public void performQuack(){
        quackBehavior.quack();
    }

    public void setQuackBehavior(QuackBehavior quackBehavior){
        this.quackBehavior = quackBehavior;
    }

    public void swim() {
        System.out.println("All ducks float, even decoys!");
    }
}

File: Strategy/TypeScript/src/duck/Duck.ts
import { FlyBehavior } from '../fly/FlyBehavior';
import { QuackBehavior } from '../quack/QuackBehavior';

class Duck {
  flyBehavior: FlyBehavior;
  quackBehavior: QuackBehavior;

  display() {
    throw new Error('Not implemented');
  }

  performFly() {
    this.flyBehavior.fly();
  }

  setFlyBehavior(flyBehavior: FlyBehavior) {
    this.flyBehavior = flyBehavior;
  }

  performQuack() {
    this.quackBehavior.quack();
  }

  setQuackBehavior(quackBehavior: QuackBehavior) {
    this.quackBehavior = quackBehavior;
  }

  swim() {
    console.log('All ducks float, even decoys!');
  }
}

export default Duck;
